---
title: "Class 06: R Functions"
author: "Matthew Chan (PID: A18130675)"
format: html

---

## Background

Functions are at the heart of using R. Everything we do involves calling an dusing functions (from data input, analysis to result output)

All functions in R have at least 3 things: 

1. A **name** the thing we use to call then function
2. One or more input **arguments**
3. The **body**, lines of code between curly brackets { } that does the work of the function.

## A first function 


Let's write a silly function to add numbers
```{r}
add <- function (x) {
  x + 1 
  
}
```

Lets try it out
```{r}
add(100)
```
Will this work
```{r}
add (c(100,200,300))
```

Modify to be more useful and add more than just 1

```{r}
add <- function (x, y=1) {
  x + y
  
}
```

```{r}
add(100, 10)
```
Will this work?
```{r}
add(100,)
```
 
```{r}
plot (1:10, col="blue", typ="b")
```

```{r}
log(10, base=10)
```

> **Note** Input arguments can either be **required** or **optional**, optional is specified with an equal sign as a default option


```{r}
#add(x=100, y=200, z=300)
```


## A second function

All functions in R look like this

```
name <- function(argument) {
  body
}
```

The `sample()` function in R...

```{r}
sample(1:10, size=4)
```
>Q. Return 12 numbers picked randomly from the input 1:10

```{r}
sample(1:10, size=12, replace=TRUE)
```


>Q. Write the code to generate a 12 nucleotide long DNA sequence

```{r}
bases <- c("A","G","C","T")
DNAseq <- sample(bases, size=12, replace=T)
  paste(DNAseq, collapse="")
```

>Q. Write a first version function called `generate_dna()` that generates a user specified length 'n' random DNA sequence?

```{r}
generate_dna <- function(x=100) {
  bases <- c("A","G","C","T")
  DNAseq <- sample(bases, size=x, replace=T)
  ans <- paste(DNAseq, collapse="")
  return(ans) 
}
```

```{r}
generate_dna(200)
```

```{r}
x<- c("H", "E", "L", "L", "O")
paste(x, collapse="hello")
```

>Q. Give the user an option to return FASTA format output sequence or standard multi-element vector format

```{r}
generate_dna <- function(x=100, fasta=T) {
  bases <- c("A","G","C","T")
  DNAseq <- sample(bases, size=x, replace=T)
  
  if(fasta) {
    DNAseq <- paste(DNAseq, collapse="")
      cat("Bye")
  } else {
      cat("Hello")
    }
return(DNAseq)
}
```

```{r}
generate_dna(5, fasta=F)
```

## A new cool function
>Q. Write a function called `generate_protein` that generates a user specified length protein sequence in FASTA like format?


```{r}
generate_protein <- function(x=6) {
  aminoacids <- c("A", "R", "N", "D", "C", "Q", "E", "G", "H", "I", "L", "K", "M", "F", "P", "S", "T", "W", "Y", "V")
  

  protein <- sample(aminoacids, size = x, replace = T)
  protein <- paste(protein, collapse="")
  return(protein)
}
```


>Q. Use your new `generate_protein` function to generate all sequences between length 6 and 12 amino-acids in length and check if any of these are unique in nature (ie found in the NR database at NCBI)?

This is the wrong code but randomly generates protein sequences between the lengths of 6 and 12 amino acids
```{r}

generate_protein <- function(x=6) {
  aminoacids <- c("A", "R", "N", "D", "C", "Q", "E", "G", "H", "I", "L", "K", "M", "F", "P", "S", "T", "W", "Y", "V")
  length <- sample(6:12,size = 1, replace=T)

  protein <- sample(aminoacids, size = length, replace = T)
  protein <- paste(protein, collapse="")
  
  return(protein)
}
```

```{r}
generate_protein()
```
VEGWCAEC matches with Accession Number 	NJD86855.1

```{r}
generate_protein(6)
generate_protein(7)
generate_protein(8)
generate_protein(9)
generate_protein(10)
generate_protein(11)
generate_protein(12)
```


Or we could fo a `for()` loop:

```{r}
for (i in 6:12) {
  cat(">", sep="", i, "\n")
  
  cat(generate_protein(i), "\n" )
}
```

